src/test/java/com/google/sps/DetectTextFromImageTest.java

```diff
@@ -0,0 +1,150 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps;
+
+import com.google.cloud.vision.v1.AnnotateImageRequest;
+import com.google.cloud.vision.v1.ImageSource;
+import com.google.cloud.vision.v1.Feature;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.jupiter.api.Assertions;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Test for DetectTextFromImage class. */
+@RunWith(JUnit4.class)
+public final class DetectTextFromImageTest {
+
+  private final String IMAGE_BASE_URI =
+      "https://shop-by-photos-step-2020.ey.r.appspot.com/get-image-url?blob-key=";
+  private final Feature TEXT_DETECTION_FEATURE = Feature.newBuilder().setType(Feature.Type.TEXT_DETECTION).build();
+
+  private String nullImageKey = "";
+  private String invalidImageKey = "abc £$%  d\n\n\n";
+  // This key is copied from Cloud Datastore. Make sure it exists there before running the tests.
+  private String validImageKey =
+      "AMIfv943BzZrlhyLIN_el9l15Zz2LUp4H3bRNwHj2yvPaDNuv0uZjsp8vSySw0u5pEp-sspSVv4U5fkCT6SY1vbTeObItcslXvPswVieK3rInnCc0nBkTDlfIgWqxlvYjFVPS1QQqCjiY7NlRNSHA3gRXtsv6hrj7_J3_c_DaCP4jhpbZid2bkKvOA1XD0geHOXbCjKHf0ZIbEU4wKqQgGcCxT0X4Ddi4o6Nk8rhN7CNbN27GD-iBfrIG2RwNXr24xLLlFg8Bw9xCUxrbaXO6tkJNHndSghb0x4KAZ-IEPYKHfuWtLfXSmgMV_D1hHZdqShTaUZIyJu-";
+
+  private List<String> nullShoppingList = new ArrayList<>();
+  private List<String> validShoppingList = new ArrayList<>();
+  private List<String> invalidShoppingList = new ArrayList<>();
+
+  private DetectTextFromImage detectText;
+
+  @Before
+  public void setUp() {
+    validShoppingList.add("Text:Bag ");
+    validShoppingList.add(
+        "Position : vertices { x: 151 y: 94 } vertices { x: 187 y: 94 } vertices { x: 187 y: 115 } vertices { x: 151 y: 115 }");
+    validShoppingList.add("Text:Bag");
+    validShoppingList.add(
+        "Position : vertices { x: 152 y: 94 } vertices { x: 187 y: 96 } vertices { x: 186 y: 114 } vertices { x: 151 y: 112 }");
+
+    invalidShoppingList.add("!%^&*+");
+  }
+
+  /**
+   * ********************************************** 
+   * **** TESTS FOR DetectTextFromImage constructor
+   * **********************************************
+   */
+  @Test
+  public void DetectTextFromImage_forNullImageKey() {
+    Assertions.assertThrows(
+        PhotoShoppingException.class,
+        () -> {
+            new DetectTextFromImage(nullImageKey);
+        });
+  }
+
+  @Test
+  public void DetectTextFromImage_forInvalidImageKey() {
+    Assertions.assertThrows(
+        PhotoShoppingException.class,
+        () -> {
+            new DetectTextFromImage(invalidImageKey);
+        });
+  }
+
+  /**
+   * ******************************************** 
+   * **** TESTS FOR shoppingImageRequestGenerator
+   * ********************************************
+   */
```

Comment:
I think you only need to test one method:
ImageTextDetector#detect

Those tests can be done in this way:

Define an interface for the CloudVisionAPI so you will have:

FakeCloudVisionAPI implements CloudVisionAPI {
 // implement the methods and add additional methods to do some setups. For example, 
//  there could be a method called "setReturnValue" which will set a return value to be used
// for all request.
}

In your ImageTextDetector, you will accept the interface, e.g. "CloudVisionAPI". In production code, you will pass in a real implementation, e.g. "CloudVisionAPIImpl" which calls the real API. For tests, you will create a Fake implementation and pass into the "ImageTextDetector" and you can use the fake implementation to fake all the behaviors you want. 

Happy to discuss more :)

https://blog.pragmatists.com/test-doubles-fakes-mocks-and-stubs-1a7491dfa3da

Reply:
Thank you Changcheng.
I'll implement this in new prs.
So now its just the ImageTextDetector class in this pr. (Also the exception class.)